# Week 3: Baseline Predictive Modeling (Booking Cancellation Prediction)

## Objective
Build a baseline machine learning model to predict hotel booking cancellations.

## Workflow
1. Load Dataset
2. Verify Dataset
3. Select Features
4. Train-Test Split
5. Train Model
6. Evaluate Model

## Libraries Used
- Pandas
- NumPy
- Scikit-Learn

In [2]:
import pandas as pd

df = pd.read_csv(
    r"C:\Users\cusac\OneDrive - Thiagarajar School of Management\Desktop\Hotel Bookings Project\Data\hotel_bookings_cleaned.csv"
)

print(df.shape)
df.head()

(84889, 35)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,total_stay_nights,total_guests,total_revenue
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,0,Transient,0.0,0,0,Check-Out,7/1/2015,0,2,0.0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,0,Transient,0.0,0,0,Check-Out,7/1/2015,0,2,0.0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,0,Transient,75.0,0,0,Check-Out,7/2/2015,1,1,75.0
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,0,Transient,75.0,0,0,Check-Out,7/2/2015,1,1,75.0
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,0,Transient,98.0,0,1,Check-Out,7/3/2015,2,2,196.0


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84889 entries, 0 to 84888
Data columns (total 35 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   hotel                           84889 non-null  object 
 1   is_canceled                     84889 non-null  int64  
 2   lead_time                       84889 non-null  int64  
 3   arrival_date_year               84889 non-null  int64  
 4   arrival_date_month              84889 non-null  object 
 5   arrival_date_week_number        84889 non-null  int64  
 6   arrival_date_day_of_month       84889 non-null  int64  
 7   stays_in_weekend_nights         84889 non-null  int64  
 8   stays_in_week_nights            84889 non-null  int64  
 9   adults                          84889 non-null  int64  
 10  children                        84889 non-null  int64  
 11  babies                          84889 non-null  int64  
 12  meal                            

In [4]:
df.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                         

In [5]:
df.describe()

,is_canceled,lead_time,arrival_date_year,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,...,booking_changes,agent,company,days_in_waiting_list,adr,required_car_parking_spaces,total_of_special_requests,total_stay_nights,total_guests,total_revenue
count,84889.000000,84889.000000,84889.000000,84889.000000,84889.000000,84889.000000,84889.00000,84889.000000,84889.000000,84889.000000,...,84889.000000,84889.000000,84889.000000,84889.000000,84889.000000,84889.000000,84889.000000,84889.000000,84889.000000,84889.000000
mean,0.271637,80.150444,2016.199637,26.721907,15.798337,1.001319,2.61313,1.866755,0.117683,0.010472,...,0.268822,78.846258,11.307001,0.768250,101.788429,0.082013,0.693541,3.614449,1.994911,373.848770
std,0.444806,86.460475,0.685912,13.812371,8.842121,1.032651,2.05764,0.627788,0.416169,0.112255,...,0.724677,109.037505,54.687341,10.145598,45.205950,0.278223,0.827720,2.767990,0.768068,334.516873
min,0.000000,0.000000,2015.000000,1.000000,1.000000,0.000000,0.00000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,-6.380000,0.000000,0.000000,0.000000,0.000000,-63.800000
25%,0.000000,11.000000,2016.000000,15.000000,8.000000,0.000000,1.00000,2.000000,0.000000,0.000000,...,0.000000,9.000000,0.000000,0.000000,71.100000,0.000000,0.000000,2.000000,2.000000,150.000000
50%,0.000000,49.000000,2016.000000,27.000000,16.000000,1.000000,2.00000,2.000000,0.000000,0.000000,...,0.000000,9.000000,0.000000,0.000000,96.530000,0.000000,0.000000,3.000000,2.000000,291.600000
75%,1.000000,125.000000,2017.000000,37.000000,23.000000,2.000000,3.00000,2.000000,0.000000,0.000000,...,0.000000,195.000000,0.000000,0.000000,130.000000,0.000000,1.000000,5.000000,2.000000,484.000000
max,1.000000,737.000000,2017.000000,53.000000,31.000000,19.000000,50.00000,55.000000,10.000000,10.000000,...,21.000000,535.000000,543.000000,391.000000,227.000000,8.000000,5.000000,69.000000,55.000000,7590.000000


## Dataset Verification

- Dataset loaded successfully.
- Total Records: 84,889
- Total Features: 35
- No missing values found.
- Dataset is ready for predictive modeling.

# Week 3: Baseline Predictive Modeling

## Objective

## Workflow

## Libraries Used

Import Libraries

Load Dataset

Display First 5 Records

Dataset Shape

Dataset Information

Column Names

Missing Values

Statistical Summary

Dataset Verification

## Step 1: Feature Selection

In this step, relevant features are selected for building the baseline booking cancellation prediction model. The target variable is `is_canceled`.

In [6]:
# Features
X = df.drop(columns=['is_canceled'])

# Target
y = df['is_canceled']

print("Feature Matrix Shape:", X.shape)
print("Target Shape:", y.shape)

Feature Matrix Shape: (84889, 34)
Target Shape: (84889,)


In [7]:
y.value_counts()

is_canceled
0    61830
1    23059
Name: count, dtype: int64

In [8]:
y.value_counts(normalize=True) * 100

is_canceled
0    72.836292
1    27.163708
Name: proportion, dtype: float64

In [9]:
#Identifies categorial and numerical columns
categorical_cols = X.select_dtypes(include=['object']).columns
numerical_cols = X.select_dtypes(exclude=['object']).columns

print("Categorical Features:", len(categorical_cols))
print("Numerical Features:", len(numerical_cols))

Categorical Features: 12
Numerical Features: 22


In [10]:
#Display column names
print("Categorical Columns:")
print(list(categorical_cols))

print("\nNumerical Columns:")
print(list(numerical_cols))

Categorical Columns:
['hotel', 'arrival_date_month', 'meal', 'country', 'market_segment', 'distribution_channel', 'reserved_room_type', 'assigned_room_type', 'deposit_type', 'customer_type', 'reservation_status', 'reservation_status_date']

Numerical Columns:
['lead_time', 'arrival_date_year', 'arrival_date_week_number', 'arrival_date_day_of_month', 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'is_repeated_guest', 'previous_cancellations', 'previous_bookings_not_canceled', 'booking_changes', 'agent', 'company', 'days_in_waiting_list', 'adr', 'required_car_parking_spaces', 'total_of_special_requests', 'total_stay_nights', 'total_guests', 'total_revenue']


## Feature Selection Summary

The dataset has been prepared for predictive modeling.

- Target Variable: is_canceled
- Total Features: 34
- Numerical Features: 22
- Categorical Features: 12

The dataset is now ready for preprocessing and train-test splitting in the next stage of the machine learning pipeline.

In [11]:
X_encoded = pd.get_dummies(X, drop_first=True)

print("Encoded Feature Matrix Shape:", X_encoded.shape)

Encoded Feature Matrix Shape: (84889, 1174)


In [12]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X_encoded,
    y,
    test_size=0.2,
    random_state=42
)

In [13]:
print("Training Features:", X_train.shape)
print("Testing Features :", X_test.shape)
print("Training Target  :", y_train.shape)
print("Testing Target   :", y_test.shape)

Training Features: (67911, 1174)
Testing Features : (16978, 1174)
Training Target  : (67911,)
Testing Target   : (16978,)


## Train-Test Split

The encoded dataset has been divided into training and testing sets using an 80:20 ratio.

- Training Data: 67,911 records
- Testing Data: 16,978 records

The training dataset will be used to train the machine learning model, while the testing dataset will be used to evaluate the model's performance on unseen data.